src/metabase/models/collection.clj

```diff
@@ -409,39 +411,55 @@
                          [(mdb/unique-identifier) user-id])}
    (fn user->personal-collection-id*
      [user-id]
-     (u/the-id (user->personal-collection user-id)))
+     (some-> user-id user->personal-collection u/the-id))
    ;; cache the results for 60 minutes; TTL is here only to eventually clear out old entries/keep it from growing too
    ;; large
    :ttl/threshold (* 60 60 1000)))
 
-(mu/defn user->personal-collection-and-descendant-ids :- [:sequential {:min 1} ms/PositiveInt]
+(mu/defn user->personal-collection-and-descendant-ids :- [:maybe [:sequential {:min 1} ms/PositiveInt]]
   "Somewhat-optimized function that fetches the ID of a User's Personal Collection as well as the IDs of all descendants
   of that Collection. Exists because this needs to be known to calculate the Current User's permissions set, which is
   done for every API call; this function is an attempt to make fetching this information as efficient as reasonably
   possible."
   [user-or-id]
-  (let [personal-collection-id (user->personal-collection-id (u/the-id user-or-id))]
+  (when-let [personal-collection-id (user->personal-collection-id (u/the-id user-or-id))]
     (cons personal-collection-id
           ;; `descendant-ids` wants a CollectionWithLocationAndID, and luckily we know Personal Collections always go
           ;; in Root, so we can pass it what it needs without actually having to fetch an entire CollectionInstance
           (descendant-ids {:location "/", :id personal-collection-id}))))
 
+(mu/defn user->existing-personal-collection-and-descendant-ids :- [:maybe [:sequential {:min 1} ms/PositiveInt]]
```

Comment:
I'm not sure if I'm understanding this function properly, or why it's needed. It's the same as `user->personal-collection-and-descendant-ids` but it doesn't create the collection if it doesn't exist? Is there actually a difference in caching (per the docstring) or is that mistaken?

Reply:
Oops, yes, thank you - I actually initially created this function thinking we would solve the problem by making the caller decide whether to create the personal collection or just fetch an existing one, but then later decided it was better to just check to see whether the user was an API key user before creating the collection. So now we can just always use `user->personal-collection-and-descendant-ids`, which will create personal collections for normal users (as we did before) but return an empty coll for API key users.